tempesta_fw/http_sess.c

```diff
@@ -0,0 +1,731 @@
+	 */
+	r = tfw_http_sticky_get(req, &cookie_val);
+	if (r < 0)
+		return r;
+	if (r == 0)
+		return tfw_http_sticky_notfound(req);
+	if (r == 1) {
+		/*
+		 * Verify sticky cookie value: if it's wrong, then this can be
+		 * an attack as well as changed Tempesta or whatever else.
+		 * The first case must be properly handled by Frang limit
+		 * (TODO #598), your ever can limit number of invalid sticky
+		 * cookie tries to 1. While we just send normal 302 redirect to
+		 * keep user experience intact.
+		 */
+		if (tfw_http_sticky_verify(req, &cookie_val, sv))
+			return tfw_http_sticky_send_302(req, sv) ? : 1;
+		return 0;
+	}
+	TFW_WARN("Multiple Tempesta sticky cookies found: %d\n", r);
+
+	return -1;
+}
+
+/*
+ * Add Tempesta sticky cookie to an HTTP response if needed.
+ */
+int
+tfw_http_sess_resp_process(TfwHttpResp *resp, TfwHttpReq *req)
+{
+	if (!tfw_cfg_sticky.enabled)
+		return 0;
+	BUG_ON(!req->sess);
+
+	/*
+	 * RFC 6265 4.1.1 and 4.1.2 says that we should not set session cookie
+	 * if it's not necessary. Since client didn't send up the cookie and
+	 * it seems that we don't enforce them, we can just set the cookie in
+	 * each response forwarded to the client.
+	 */
+	if (req->flags & TFW_HTTP_HAS_STICKY_COOKIE)
+		return 0;
+	return tfw_http_sticky_add(resp, req);
+}
+
+void
+tfw_http_sess_put(TfwHttpSess *sess)
+{
+	if (atomic_dec_and_test(&sess->users))
+		/*
+		 * Use counter reached 0, so session already expired and evicted
+		 * from the hash table.
+		 */
+		kmem_cache_free(sess_cache, sess);
+}
+
+/**
+ * Obtains appropriate HTTP session for the request based on Sticky cookies.
+ */
+int
+tfw_http_sess_obtain(TfwHttpReq *req)
+{
+	int r;
+	unsigned long key = 0, crc_tmp = 0;
+	TfwHttpSess *sess;
+	SessHashBucket *hb;
+	struct hlist_node *tmp;
+	StickyVal sv = { };
+
+	if (!tfw_cfg_sticky.enabled)
+		return 0;
+
+	if ((r = tfw_http_sticky_req_process(req, &sv)))
+		return r;
+
+	/*
+	 * Sticky cookie can be not enforced and we still have to allocate new
+	 * session for requests w/o session cookie. It means that malicious user
+	 * can always send us requests w/o session cookie. HMAC will be
+	 * different due to different ingress timestamps, so DoS is very
+	 * possible. The only thing which we can do is to enforce the cookie.
+	 * However, we cal loose innocent clients w/ disabled cookies.
+	 * We leave this for administrator decision or more progressive DDoS
+	 * mitigation techniques.
+	 */
+
+	if (!sv.ts) {
+		/* No sticky cookie in request and no enforcement. */
+		sv.ts = jiffies;
+		if (__sticky_calc(req, &sv))
+			return -1;
+	}
+
+	__tdb_hash_calc(&key, &crc_tmp, sv.hmac, sizeof(sv.hmac));
+	key |= crc_tmp << 32;
+
+	hb = &sess_hash[hash_min(key, SESS_HASH_BITS)];
+
+	spin_lock(&hb->lock);
+
+	hlist_for_each_entry_safe(sess, tmp, &hb->list, hentry) {
+		/* Collect garbage first to not to return expired session. */
+		if (sess->expires < jiffies) {
+			hash_del(&sess->hentry);
+			tfw_http_sess_put(sess);
```

Comment:
Requests may potentially stay for a long time in Tempesta due to various reasons. What if that session is still part of some request? Then the `req->sess` pointer will become stale. Is there something that I am missing here?